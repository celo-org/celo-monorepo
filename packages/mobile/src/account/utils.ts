import { ADDRESS_LENGTH } from 'src/exchange/reducer'

export const isAddressFormat = (content: string): boolean => {
  return content.startsWith('0x') && content.length === ADDRESS_LENGTH
}
