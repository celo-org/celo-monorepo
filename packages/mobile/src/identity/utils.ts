import BigNumber from 'bignumber.js'

export function isUserBalanceSufficient(
  userBalance: string | number | null,
  estimatedTxAmount: number | string
) {
  if (!userBalance || new BigNumber(userBalance).isLessThan(new BigNumber(estimatedTxAmount))) {
    return false
  }

  return true
}
