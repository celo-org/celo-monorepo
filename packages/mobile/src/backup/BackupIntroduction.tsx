import Button, { BtnTypes } from '@celo/react-components/components/Button'
import colors from '@celo/react-components/styles/colors'
import { fontStyles } from '@celo/react-components/styles/fonts'
import * as React from 'react'
import { WithNamespaces, withNamespaces } from 'react-i18next'
import { ActivityIndicator, Image, ScrollView, StyleSheet, Text, View } from 'react-native'
import SafeAreaView from 'react-native-safe-area-view'
import { connect } from 'react-redux'
import { setBackupDelayed } from 'src/account/actions'
import CeloAnalytics from 'src/analytics/CeloAnalytics'
import { CustomEventNames } from 'src/analytics/constants'
import componentWithAnalytics from 'src/analytics/wrapper'
import { enterBackupFlow, exitBackupFlow, navigatePinProtected } from 'src/app/actions'
import { Namespaces } from 'src/i18n'
import backupIcon from 'src/images/backup-icon.png'
import { headerWithBackButton } from 'src/navigator/Headers'
import { navigate, navigateBack } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import { RootState } from 'src/redux/reducers'
import { isBackupTooLate } from 'src/redux/selectors'

interface StateProps {
  backupCompleted: boolean
  socialBackupCompleted: boolean
  backupTooLate: boolean
  backupDelayedTime: number
  doingPinVerification: boolean
}

interface DispatchProps {
  setBackupDelayed: typeof setBackupDelayed
  enterBackupFlow: typeof enterBackupFlow
  exitBackupFlow: typeof exitBackupFlow
  navigatePinProtected: typeof navigatePinProtected
}

type Props = WithNamespaces & StateProps & DispatchProps

const mapStateToProps = (state: RootState): StateProps => {
  return {
    backupCompleted: state.account.backupCompleted,
    socialBackupCompleted: state.account.socialBackupCompleted,
    backupTooLate: isBackupTooLate(state),
    backupDelayedTime: state.account.backupDelayedTime,
    doingPinVerification: state.app.doingPinVerification,
  }
}

class BackupIntroduction extends React.Component<Props> {
  static navigationOptions = () => ({
    ...headerWithBackButton,
  })

  componentDidMount() {
    this.props.enterBackupFlow()
  }

  componentWillUnmount() {
    this.props.exitBackupFlow()
  }

  onPressViewBackupKey = () => {
    CeloAnalytics.track(CustomEventNames.view_backup_phrase)
    this.props.navigatePinProtected(Screens.BackupPhrase)
  }

  onPressBackup = () => {
    CeloAnalytics.track(CustomEventNames.set_backup_phrase)
    this.props.navigatePinProtected(Screens.BackupPhrase)
  }

  onPressSetupSocialBackup = () => {
    CeloAnalytics.track(CustomEventNames.set_social_backup)
    navigate(Screens.BackupSocialIntro)
  }

  onPressViewSocialBackup = () => {
    CeloAnalytics.track(CustomEventNames.view_social_backup)
    this.props.navigatePinProtected(Screens.BackupSocial)
  }

  onPressDelay = () => {
    this.props.setBackupDelayed()
    CeloAnalytics.track(CustomEventNames.delay_backup)
    navigateBack()
  }

  render() {
    const {
      t,
      backupDelayedTime,
      backupTooLate,
      backupCompleted,
      socialBackupCompleted,
      doingPinVerification,
    } = this.props
    return (
      <SafeAreaView style={styles.container}>
        <ScrollView contentContainerStyle={styles.scrollContainer}>
          <Image source={backupIcon} style={styles.logo} />
          <Text style={styles.h1}>{t('backupAndRecovery')}</Text>
          {!backupCompleted && (
            <>
              <Text style={styles.body}>{t('backupKeyIntro.0')}</Text>
              <Text style={[styles.body, fontStyles.bold]}>{t('backupKeyIntro.1')}</Text>
            </>
          )}
          {backupCompleted &&
            !socialBackupCompleted && (
              <>
                <Text style={styles.body}>
                  {t('backupKeyIntro.2')}
                  <Text style={[styles.body, fontStyles.bold]}>{t('backupKeyIntro.3')}</Text>
                </Text>
                <Text style={styles.body}>{t('backupKeyIntro.4')}</Text>
              </>
            )}
          {backupCompleted &&
            socialBackupCompleted && (
              <>
                <Text style={styles.body}>
                  {t('backupKeyIntro.2')}
                  <Text style={[styles.body, fontStyles.bold]}>{t('backupKeyIntro.3')}</Text>
                </Text>
                <Text style={styles.body}>
                  {t('backupKeyIntro.5')}
                  <Text style={[styles.body, fontStyles.bold]}>{t('backupKeyIntro.6')}</Text>
                </Text>
              </>
            )}
        </ScrollView>
        {doingPinVerification && (
          <ActivityIndicator size="large" color={colors.celoGreen} style={styles.loader} />
        )}
        <View>
          {!backupCompleted && (
            <>
              <Button
                onPress={this.onPressBackup}
                text={t('getYourKey')}
                standard={false}
                type={BtnTypes.PRIMARY}
              />
              {backupTooLate &&
                !backupDelayedTime && (
                  <Button
                    onPress={this.onPressDelay}
                    text={t('delayBackup')}
                    standard={false}
                    type={BtnTypes.SECONDARY}
                  />
                )}
            </>
          )}

          {backupCompleted &&
            !socialBackupCompleted && (
              <>
                <Button
                  onPress={this.onPressSetupSocialBackup}
                  text={t('setUpSocialBackup')}
                  standard={false}
                  type={BtnTypes.PRIMARY}
                />
                <Button
                  onPress={this.onPressViewBackupKey}
                  text={t('viewBackupKey')}
                  standard={false}
                  type={BtnTypes.SECONDARY}
                />
              </>
            )}

          {backupCompleted &&
            socialBackupCompleted && (
              <>
                <Button
                  onPress={this.onPressBackup}
                  text={t('viewBackupKey')}
                  standard={false}
                  type={BtnTypes.SECONDARY}
                />
                <Button
                  onPress={this.onPressViewSocialBackup}
                  text={t('viewSafeguards')}
                  standard={false}
                  type={BtnTypes.SECONDARY}
                />
              </>
            )}
        </View>
      </SafeAreaView>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: colors.background,
    justifyContent: 'space-between',
  },
  scrollContainer: {
    flex: 1,
    paddingHorizontal: 30,
    paddingBottom: 20,
    justifyContent: 'center',
  },
  logo: {
    alignSelf: 'center',
    height: 75,
    width: 75,
  },
  h1: {
    ...fontStyles.h1,
    marginTop: 15,
  },
  body: {
    ...fontStyles.body,
    textAlign: 'center',
    paddingBottom: 15,
  },
  loader: {
    marginBottom: 20,
  },
})

export default componentWithAnalytics(
  connect<StateProps, DispatchProps, {}, RootState>(
    mapStateToProps,
    {
      setBackupDelayed,
      enterBackupFlow,
      exitBackupFlow,
      navigatePinProtected,
    }
  )(withNamespaces(Namespaces.backupKeyFlow6)(BackupIntroduction))
)
