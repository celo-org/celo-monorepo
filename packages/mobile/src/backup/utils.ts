import * as _ from 'lodash'
import { generateMnemonic, wordlists } from 'react-native-bip39'
import { getKey } from 'src/utils/keyStore'
import Logger from 'src/utils/Logger'

const TAG = 'Backup/utils'

export const DAYS_TO_BACKUP = 1
export const DAYS_TO_DELAY = 1 / 24 // 1 hour delay
export const MNEMONIC_SPLITTER = 'celo'

export async function createQuizWordList(mnemonic: string, language: string | null) {
  const disallowedWordSet = new Set(mnemonic.split(' '))
  const languageWordList = getWordlist(language)
  const wordOptions: string = await generateMnemonic(1000, null, languageWordList)
  return wordOptions.split(' ').filter((word: string) => !disallowedWordSet.has(word))
}

export function selectQuizWordOptions(
  mnemonic: string,
  allWords: string[],
  numOptions: number
): [string, string[]] | [] {
  const correctWord = _.sample(mnemonic.split(' '))

  if (!correctWord) {
    // mnemonic is empty
    return []
  }

  const wordOptions = _.chain(allWords)
    .sampleSize(numOptions - 1)
    .push(correctWord)
    .shuffle()
    .value()

  return [correctWord, wordOptions]
}

export function getWordlist(language: string | null) {
  if (!language) {
    return wordlists.EN
  }

  switch (language.slice(0, 2)) {
    case 'es': {
      return wordlists.ES
    }
    default: {
      return wordlists.EN
    }
  }
}

// Split a mnemonic into two and insert the mnemonic splitter in between
export function splitMnemonic(mnemonic: string, language: string | null): string[] {
  // TODO use language to i18n the splitter word? For now just using 'celo' everywhere

  if (!mnemonic) {
    throw new Error('Cannot split invalid mnemonic')
  }

  const mnemonicWords = mnemonic.split(' ')
  const firstHalf = [...mnemonicWords.slice(0, mnemonicWords.length / 2), MNEMONIC_SPLITTER]
  const secondHalf = [MNEMONIC_SPLITTER, ...mnemonicWords.slice(mnemonicWords.length / 2)]
  return [firstHalf.join(' '), secondHalf.join(' ')]
}

export function joinMnemonic(mnemonicShards: string[]) {
  if (
    !mnemonicShards ||
    !(mnemonicShards.length === 2) ||
    !mnemonicShards[0].includes(MNEMONIC_SPLITTER) ||
    !mnemonicShards[1].includes(MNEMONIC_SPLITTER)
  ) {
    throw new Error('Cannot join invalid mnemonic shards')
  }

  if (mnemonicShards[0].startsWith(MNEMONIC_SPLITTER)) {
    mnemonicShards.reverse()
  }

  const [firstHalf, secondHalf] = mnemonicShards.map((shard) => shard.split(' '))
  return [...firstHalf.slice(0, firstHalf.length - 1), ...secondHalf.slice(1)].join(' ')
}

export async function getStoredMnemonic(): Promise<string | null> {
  try {
    Logger.debug(TAG, 'Checking keystore for mnemonic')
    const mnemonic = await getKey('mnemonic')

    if (!mnemonic) {
      throw new Error('No mnemonic found in storage')
    }

    return mnemonic
  } catch (error) {
    Logger.error(TAG, 'Failed to retrieve mnemonic', error)
    return null
  }
}
