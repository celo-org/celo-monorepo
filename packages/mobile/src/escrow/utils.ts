import { ensureLeading0x } from '@celo/base'
import { generateDeterministicInviteCode } from '@celo/utils/lib/account'
import { publicKeyToAddress } from '@celo/utils/lib/address'

export const generateEscrowPaymentIdAndPk = (
  recipientPhoneHash: string,
  recipientPepper: string,
  addressIndex: number = 0
) => {
  const { publicKey, privateKey } = generateDeterministicInviteCode(
    recipientPhoneHash,
    recipientPepper,
    addressIndex
  )
  return { paymentId: publicKeyToAddress(publicKey), privateKey: ensureLeading0x(privateKey) }
}

export const generateUniquePaymentId = (
  existingPaymentIds: string[],
  phoneHash: string,
  pepper: string
) => {
  const paymentIdSet: Set<string> = new Set(existingPaymentIds)

  // Using an upper bound of 100 to be sure this doesn't run forever given
  // the realistic amount of pending escrow txs is far less than this
  for (let i = 0; i < 100; i += 1) {
    const { paymentId } = generateEscrowPaymentIdAndPk(phoneHash, pepper, i)
    if (!paymentIdSet.has(paymentId)) {
      return paymentId
    }
  }
}
