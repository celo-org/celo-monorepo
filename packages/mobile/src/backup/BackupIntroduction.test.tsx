import * as React from 'react'
import 'react-native'
import { Provider } from 'react-redux'
import * as renderer from 'react-test-renderer'
import BackupIntroduction from 'src/backup/BackupIntroduction'
import { createMockStore } from 'test/utils'

const NOW = new Date().getTime()

describe('BackupIntroduction', () => {
  it('renders correctly when neither backup nor social backup are complete', () => {
    const tree = renderer.create(
      <Provider store={createMockStore({})}>
        <BackupIntroduction />
      </Provider>
    )
    expect(tree).toMatchSnapshot()
  })

  it('renders correctly when backup completed and social backup not completed', () => {
    const tree = renderer.create(
      <Provider
        store={createMockStore({
          account: { backupCompleted: true, socialBackupCompleted: false },
        })}
      >
        <BackupIntroduction />
      </Provider>
    )
    expect(tree).toMatchSnapshot()
  })

  it('renders correctly when backup completed and social backup completed', () => {
    const tree = renderer.create(
      <Provider
        store={createMockStore({
          account: { backupCompleted: true, socialBackupCompleted: true },
        })}
      >
        <BackupIntroduction />
      </Provider>
    )
    expect(tree).toMatchSnapshot()
  })

  it('renders correctly when backup too late and no delay', () => {
    const tree = renderer.create(
      <Provider
        store={createMockStore({
          account: { accountCreationTime: NOW - 1000 },
        })}
      >
        <BackupIntroduction />
      </Provider>
    )
    expect(tree).toMatchSnapshot()
  })

  it('renders correctly when backup too late and delayed', () => {
    const tree = renderer.create(
      <Provider
        store={createMockStore({
          account: { accountCreationTime: NOW - 1000, backupDelayedTime: NOW },
        })}
      >
        <BackupIntroduction />
      </Provider>
    )
    expect(tree).toMatchSnapshot()
  })
  it('renders correctly when pin verification is in-progress', () => {
    const tree = renderer.create(
      <Provider
        store={createMockStore({
          app: { doingPinVerification: true },
        })}
      >
        <BackupIntroduction />
      </Provider>
    )
    expect(tree).toMatchSnapshot()
  })
})
