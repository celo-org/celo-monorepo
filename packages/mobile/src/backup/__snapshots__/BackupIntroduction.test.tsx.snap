// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BackupIntroduction renders correctly when backup completed and social backup completed 1`] = `
<SafeAreaView
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "flex": 1,
      "justifyContent": "space-between",
    }
  }
>
  <RCTScrollView
    contentContainerStyle={
      Object {
        "flex": 1,
        "justifyContent": "center",
        "paddingBottom": 20,
        "paddingHorizontal": 30,
      }
    }
  >
    <View>
      <Image
        source={
          Object {
            "testUri": "../../../src/images/backup-icon.png",
          }
        }
        style={
          Object {
            "alignSelf": "center",
            "height": 75,
            "width": 75,
          }
        }
      />
      <Text
        style={
          Object {
            "color": "#2E3338",
            "fontFamily": "Hind-Light",
            "fontSize": 22,
            "marginTop": 15,
            "paddingBottom": 20,
            "textAlign": "center",
          }
        }
      >
        backupAndRecovery
      </Text>
      <Text
        style={
          Object {
            "color": "#2E3338",
            "fontFamily": "Hind-Regular",
            "fontSize": 16,
            "lineHeight": 24,
            "paddingBottom": 15,
            "textAlign": "center",
          }
        }
      >
        backupKeyIntro.2
        <Text
          style={
            Array [
              Object {
                "color": "#2E3338",
                "fontFamily": "Hind-Regular",
                "fontSize": 16,
                "lineHeight": 24,
                "paddingBottom": 15,
                "textAlign": "center",
              },
              Object {
                "fontFamily": "Hind-Bold",
              },
            ]
          }
        >
          backupKeyIntro.3
        </Text>
      </Text>
      <Text
        style={
          Object {
            "color": "#2E3338",
            "fontFamily": "Hind-Regular",
            "fontSize": 16,
            "lineHeight": 24,
            "paddingBottom": 15,
            "textAlign": "center",
          }
        }
      >
        backupKeyIntro.5
        <Text
          style={
            Array [
              Object {
                "color": "#2E3338",
                "fontFamily": "Hind-Regular",
                "fontSize": 16,
                "lineHeight": 24,
                "paddingBottom": 15,
                "textAlign": "center",
              },
              Object {
                "fontFamily": "Hind-Bold",
              },
            ]
          }
        >
          backupKeyIntro.6
        </Text>
      </Text>
    </View>
  </RCTScrollView>
  <View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderRadius": 3,
            "flexDirection": "row",
          },
          null,
          null,
          undefined,
          Object {
            "backgroundColor": "transparent",
          },
        ]
      }
    >
      <View
        accessible={true}
        isTVSelectable={true}
        nativeBackgroundAndroid={
          Object {
            "attribute": "selectableItemBackground",
            "type": "ThemeAttrAndroid",
          }
        }
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
            },
            Object {
              "backgroundColor": "transparent",
            },
            Object {
              "height": 50,
            },
            Object {
              "borderWidth": 0,
            },
          ]
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "space-between",
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontFamily": "Hind-SemiBold",
                  "fontSize": 16,
                },
                Object {
                  "color": "#42D689",
                },
                Object {
                  "paddingLeft": 5,
                  "paddingRight": 5,
                },
              ]
            }
          >
            viewBackupKey
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderRadius": 3,
            "flexDirection": "row",
          },
          null,
          null,
          undefined,
          Object {
            "backgroundColor": "transparent",
          },
        ]
      }
    >
      <View
        accessible={true}
        isTVSelectable={true}
        nativeBackgroundAndroid={
          Object {
            "attribute": "selectableItemBackground",
            "type": "ThemeAttrAndroid",
          }
        }
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
            },
            Object {
              "backgroundColor": "transparent",
            },
            Object {
              "height": 50,
            },
            Object {
              "borderWidth": 0,
            },
          ]
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "space-between",
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontFamily": "Hind-SemiBold",
                  "fontSize": 16,
                },
                Object {
                  "color": "#42D689",
                },
                Object {
                  "paddingLeft": 5,
                  "paddingRight": 5,
                },
              ]
            }
          >
            viewSafeguards
          </Text>
        </View>
      </View>
    </View>
  </View>
</SafeAreaView>
`;

exports[`BackupIntroduction renders correctly when backup completed and social backup not completed 1`] = `
<SafeAreaView
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "flex": 1,
      "justifyContent": "space-between",
    }
  }
>
  <RCTScrollView
    contentContainerStyle={
      Object {
        "flex": 1,
        "justifyContent": "center",
        "paddingBottom": 20,
        "paddingHorizontal": 30,
      }
    }
  >
    <View>
      <Image
        source={
          Object {
            "testUri": "../../../src/images/backup-icon.png",
          }
        }
        style={
          Object {
            "alignSelf": "center",
            "height": 75,
            "width": 75,
          }
        }
      />
      <Text
        style={
          Object {
            "color": "#2E3338",
            "fontFamily": "Hind-Light",
            "fontSize": 22,
            "marginTop": 15,
            "paddingBottom": 20,
            "textAlign": "center",
          }
        }
      >
        backupAndRecovery
      </Text>
      <Text
        style={
          Object {
            "color": "#2E3338",
            "fontFamily": "Hind-Regular",
            "fontSize": 16,
            "lineHeight": 24,
            "paddingBottom": 15,
            "textAlign": "center",
          }
        }
      >
        backupKeyIntro.2
        <Text
          style={
            Array [
              Object {
                "color": "#2E3338",
                "fontFamily": "Hind-Regular",
                "fontSize": 16,
                "lineHeight": 24,
                "paddingBottom": 15,
                "textAlign": "center",
              },
              Object {
                "fontFamily": "Hind-Bold",
              },
            ]
          }
        >
          backupKeyIntro.3
        </Text>
      </Text>
      <Text
        style={
          Object {
            "color": "#2E3338",
            "fontFamily": "Hind-Regular",
            "fontSize": 16,
            "lineHeight": 24,
            "paddingBottom": 15,
            "textAlign": "center",
          }
        }
      >
        backupKeyIntro.4
      </Text>
    </View>
  </RCTScrollView>
  <View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderRadius": 3,
            "flexDirection": "row",
          },
          null,
          null,
          undefined,
          Object {
            "backgroundColor": "#42D689",
          },
        ]
      }
    >
      <View
        accessible={true}
        isTVSelectable={true}
        nativeBackgroundAndroid={
          Object {
            "attribute": "selectableItemBackground",
            "type": "ThemeAttrAndroid",
          }
        }
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
            },
            Object {
              "backgroundColor": "#42D689",
            },
            Object {
              "height": 50,
            },
            Object {
              "borderWidth": 0,
            },
          ]
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "space-between",
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontFamily": "Hind-SemiBold",
                  "fontSize": 16,
                },
                Object {
                  "color": "#FFFFFF",
                },
                Object {
                  "paddingLeft": 5,
                  "paddingRight": 5,
                },
              ]
            }
          >
            setUpSocialBackup
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderRadius": 3,
            "flexDirection": "row",
          },
          null,
          null,
          undefined,
          Object {
            "backgroundColor": "transparent",
          },
        ]
      }
    >
      <View
        accessible={true}
        isTVSelectable={true}
        nativeBackgroundAndroid={
          Object {
            "attribute": "selectableItemBackground",
            "type": "ThemeAttrAndroid",
          }
        }
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
            },
            Object {
              "backgroundColor": "transparent",
            },
            Object {
              "height": 50,
            },
            Object {
              "borderWidth": 0,
            },
          ]
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "space-between",
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontFamily": "Hind-SemiBold",
                  "fontSize": 16,
                },
                Object {
                  "color": "#42D689",
                },
                Object {
                  "paddingLeft": 5,
                  "paddingRight": 5,
                },
              ]
            }
          >
            viewBackupKey
          </Text>
        </View>
      </View>
    </View>
  </View>
</SafeAreaView>
`;

exports[`BackupIntroduction renders correctly when backup too late and delayed 1`] = `
<SafeAreaView
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "flex": 1,
      "justifyContent": "space-between",
    }
  }
>
  <RCTScrollView
    contentContainerStyle={
      Object {
        "flex": 1,
        "justifyContent": "center",
        "paddingBottom": 20,
        "paddingHorizontal": 30,
      }
    }
  >
    <View>
      <Image
        source={
          Object {
            "testUri": "../../../src/images/backup-icon.png",
          }
        }
        style={
          Object {
            "alignSelf": "center",
            "height": 75,
            "width": 75,
          }
        }
      />
      <Text
        style={
          Object {
            "color": "#2E3338",
            "fontFamily": "Hind-Light",
            "fontSize": 22,
            "marginTop": 15,
            "paddingBottom": 20,
            "textAlign": "center",
          }
        }
      >
        backupAndRecovery
      </Text>
      <Text
        style={
          Object {
            "color": "#2E3338",
            "fontFamily": "Hind-Regular",
            "fontSize": 16,
            "lineHeight": 24,
            "paddingBottom": 15,
            "textAlign": "center",
          }
        }
      >
        backupKeyIntro.0
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#2E3338",
              "fontFamily": "Hind-Regular",
              "fontSize": 16,
              "lineHeight": 24,
              "paddingBottom": 15,
              "textAlign": "center",
            },
            Object {
              "fontFamily": "Hind-Bold",
            },
          ]
        }
      >
        backupKeyIntro.1
      </Text>
    </View>
  </RCTScrollView>
  <View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderRadius": 3,
            "flexDirection": "row",
          },
          null,
          null,
          undefined,
          Object {
            "backgroundColor": "#42D689",
          },
        ]
      }
    >
      <View
        accessible={true}
        isTVSelectable={true}
        nativeBackgroundAndroid={
          Object {
            "attribute": "selectableItemBackground",
            "type": "ThemeAttrAndroid",
          }
        }
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
            },
            Object {
              "backgroundColor": "#42D689",
            },
            Object {
              "height": 50,
            },
            Object {
              "borderWidth": 0,
            },
          ]
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "space-between",
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontFamily": "Hind-SemiBold",
                  "fontSize": 16,
                },
                Object {
                  "color": "#FFFFFF",
                },
                Object {
                  "paddingLeft": 5,
                  "paddingRight": 5,
                },
              ]
            }
          >
            getYourKey
          </Text>
        </View>
      </View>
    </View>
  </View>
</SafeAreaView>
`;

exports[`BackupIntroduction renders correctly when backup too late and no delay 1`] = `
<SafeAreaView
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "flex": 1,
      "justifyContent": "space-between",
    }
  }
>
  <RCTScrollView
    contentContainerStyle={
      Object {
        "flex": 1,
        "justifyContent": "center",
        "paddingBottom": 20,
        "paddingHorizontal": 30,
      }
    }
  >
    <View>
      <Image
        source={
          Object {
            "testUri": "../../../src/images/backup-icon.png",
          }
        }
        style={
          Object {
            "alignSelf": "center",
            "height": 75,
            "width": 75,
          }
        }
      />
      <Text
        style={
          Object {
            "color": "#2E3338",
            "fontFamily": "Hind-Light",
            "fontSize": 22,
            "marginTop": 15,
            "paddingBottom": 20,
            "textAlign": "center",
          }
        }
      >
        backupAndRecovery
      </Text>
      <Text
        style={
          Object {
            "color": "#2E3338",
            "fontFamily": "Hind-Regular",
            "fontSize": 16,
            "lineHeight": 24,
            "paddingBottom": 15,
            "textAlign": "center",
          }
        }
      >
        backupKeyIntro.0
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#2E3338",
              "fontFamily": "Hind-Regular",
              "fontSize": 16,
              "lineHeight": 24,
              "paddingBottom": 15,
              "textAlign": "center",
            },
            Object {
              "fontFamily": "Hind-Bold",
            },
          ]
        }
      >
        backupKeyIntro.1
      </Text>
    </View>
  </RCTScrollView>
  <View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderRadius": 3,
            "flexDirection": "row",
          },
          null,
          null,
          undefined,
          Object {
            "backgroundColor": "#42D689",
          },
        ]
      }
    >
      <View
        accessible={true}
        isTVSelectable={true}
        nativeBackgroundAndroid={
          Object {
            "attribute": "selectableItemBackground",
            "type": "ThemeAttrAndroid",
          }
        }
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
            },
            Object {
              "backgroundColor": "#42D689",
            },
            Object {
              "height": 50,
            },
            Object {
              "borderWidth": 0,
            },
          ]
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "space-between",
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontFamily": "Hind-SemiBold",
                  "fontSize": 16,
                },
                Object {
                  "color": "#FFFFFF",
                },
                Object {
                  "paddingLeft": 5,
                  "paddingRight": 5,
                },
              ]
            }
          >
            getYourKey
          </Text>
        </View>
      </View>
    </View>
  </View>
</SafeAreaView>
`;

exports[`BackupIntroduction renders correctly when neither backup nor social backup are complete 1`] = `
<SafeAreaView
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "flex": 1,
      "justifyContent": "space-between",
    }
  }
>
  <RCTScrollView
    contentContainerStyle={
      Object {
        "flex": 1,
        "justifyContent": "center",
        "paddingBottom": 20,
        "paddingHorizontal": 30,
      }
    }
  >
    <View>
      <Image
        source={
          Object {
            "testUri": "../../../src/images/backup-icon.png",
          }
        }
        style={
          Object {
            "alignSelf": "center",
            "height": 75,
            "width": 75,
          }
        }
      />
      <Text
        style={
          Object {
            "color": "#2E3338",
            "fontFamily": "Hind-Light",
            "fontSize": 22,
            "marginTop": 15,
            "paddingBottom": 20,
            "textAlign": "center",
          }
        }
      >
        backupAndRecovery
      </Text>
      <Text
        style={
          Object {
            "color": "#2E3338",
            "fontFamily": "Hind-Regular",
            "fontSize": 16,
            "lineHeight": 24,
            "paddingBottom": 15,
            "textAlign": "center",
          }
        }
      >
        backupKeyIntro.0
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#2E3338",
              "fontFamily": "Hind-Regular",
              "fontSize": 16,
              "lineHeight": 24,
              "paddingBottom": 15,
              "textAlign": "center",
            },
            Object {
              "fontFamily": "Hind-Bold",
            },
          ]
        }
      >
        backupKeyIntro.1
      </Text>
    </View>
  </RCTScrollView>
  <View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderRadius": 3,
            "flexDirection": "row",
          },
          null,
          null,
          undefined,
          Object {
            "backgroundColor": "#42D689",
          },
        ]
      }
    >
      <View
        accessible={true}
        isTVSelectable={true}
        nativeBackgroundAndroid={
          Object {
            "attribute": "selectableItemBackground",
            "type": "ThemeAttrAndroid",
          }
        }
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
            },
            Object {
              "backgroundColor": "#42D689",
            },
            Object {
              "height": 50,
            },
            Object {
              "borderWidth": 0,
            },
          ]
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "space-between",
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontFamily": "Hind-SemiBold",
                  "fontSize": 16,
                },
                Object {
                  "color": "#FFFFFF",
                },
                Object {
                  "paddingLeft": 5,
                  "paddingRight": 5,
                },
              ]
            }
          >
            getYourKey
          </Text>
        </View>
      </View>
    </View>
  </View>
</SafeAreaView>
`;
